multicluster/controllers/multicluster/label_identity_controller.go

```diff
@@ -0,0 +1,327 @@
+// again for checking label add.
+func (r *LabelIdentityReconciler) onPodDelete(ctx context.Context, podNamespacedName string, releaseLabelMutex bool) (ctrl.Result, error) {
+	r.labelMutex.Lock()
+	if originalLabel, isCached := r.podLabelCache[podNamespacedName]; isCached {
+		if requeue := r.requeueIfLabelResExportUpdateInProgress(originalLabel); requeue {
+			r.labelMutex.Unlock()
+			return ctrl.Result{Requeue: true}, nil
+		}
+		// Check if the original label is stale.
+		if podNames, ok := r.labelToPodsCache[originalLabel]; ok && len(podNames) == 1 && podNames.Has(podNamespacedName) {
+			klog.V(2).InfoS("Label no longer exists in the cluster, deleting corresponding ResourceExport", "label", originalLabel)
+			if err := r.deleteLabelIdentityResExport(ctx, originalLabel); err != nil {
+				r.labelMutex.Unlock()
+				return ctrl.Result{}, err
+			}
+			r.deleteFromLabelCache(podNamespacedName, originalLabel)
+		} else {
+			podNames.Delete(podNamespacedName)
+			delete(r.podLabelCache, podNamespacedName)
+		}
+	}
+	// Keep labelMutex locked if label add check will immediately follow.
+	if releaseLabelMutex {
+		r.labelMutex.Unlock()
+	}
+	return ctrl.Result{}, nil
+}
+
+// onPodCreateOrUpdate updates the Pod and label identity mapping in the cache, and
+// updates label identity ResourceExport if necessary (the Pod creation/update
+// event causes a new label identity to appear in the cluster or a label identity
+// no longer present in the cluster or both).
+func (r *LabelIdentityReconciler) onPodCreateOrUpdate(ctx context.Context, podNamespacedName, currentLabel string) (ctrl.Result, error) {
+	if result, err := r.onPodDelete(ctx, podNamespacedName, false); err != nil || result.Requeue {
+		// labelMuetx is already released.
+		return result, err
+	}
+	// If the previous part of the function did not return, the labelMutex is still
+	// locked at this point.
+	if requeue := r.requeueIfLabelResExportUpdateInProgress(currentLabel); requeue {
+		r.labelMutex.Unlock()
+		return ctrl.Result{Requeue: true}, nil
+	}
+	defer r.labelMutex.Unlock()
+	// Check if ResourceExport creation is needed for the updated label.
+	podNames, ok := r.labelToPodsCache[currentLabel]
+	if ok {
+		// This is a seen label and there's no other worker trying to delete the ResourceExport
+		// of the label concurrently. Update the cache immediately.
+		podNames.Insert(podNamespacedName)
+		r.podLabelCache[podNamespacedName] = currentLabel
+	} else {
+		// Create a ResourceExport for this label as this is a new label.
+		klog.V(2).InfoS("Creating ResourceExport for label", "label", currentLabel)
+		if err := r.createLabelIdentityResExport(ctx, currentLabel); err != nil {
+			return ctrl.Result{}, err
+		}
+		r.updateLabelCache(podNamespacedName, currentLabel)
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *LabelIdentityReconciler) requeueIfLabelResExportUpdateInProgress(label string) bool {
+	return r.labelAddsInProgress.Has(label) || r.labelDeletesInProgress.Has(label)
```

Comment:
In theory, we just need to prevent add and delete being handled the same time, but adds or deletes can happen concurrently.  But if we want to make it simpler to allow only one add/delete, I think we just need a single map like `labelExportUpateInProgress`.

Reply:
Done